README.md
# aychip

Emulator for the AY-3-8912 sound chip

## Installation

```
npm install aychip
```

## Usage

This package exports a class `AYChip` that is capable of generating audio waveforms in response to register writes. The constructor accepts an options object with the following properties:

* `frequency` (required) - The clock speed of the chip, in Hz.
* `sampleRate` (required) - The sample rate of the generated audio, in Hz.
* `volume` (optional) - The volume of the generated audio, as a number between 0 and 1. Defaults to 1.
* `stereoMode` (optional) - one of 'ABC', 'ACB', 'BAC', 'BCA', 'CAB', 'CBA' (case insensitive) indicating the positions to which each of the three AY channels should be panned. For example, 'ACB' means that channel A will be panned 50% left, channel C to the centre, and channel B 50% right. Any other value will be treated as all channels centred.
* `panning` (optional, takes precedence over `stereoMode`) - a list of three values from 0 to 1 indicating how far left or right each AY channel should be panned, with 0 being 100% left, 0.5 being centred, and 1 being 100% right. If neither `stereoMode` nor `panning` is specified, all channels will be centred.

The `AYChip` object provides the following methods:

* `setRegister(reg, val)` - Set the value of a register. `reg` is a number from 0 to 13, and `val` is a number from 0 to 255.
* `reset()` - Reset all registers to 0.
* `generate(outputBuffer, offset, count)` - Generate `count` samples of audio, writing them into `outputBuffer` - a two-element array of `Float32Array`s for the left and right channels, of sufficient length to hold the requested number of samples. Samples will be written starting at `offset`.
* `setPanning(panning)` - Set the panning for all channels. `panning` is a list of three values from 0 to 1 indicating how far left or right each AY channel should be panned, with 0 being 100% left, 0.5 being centred, and 1 being 100% right.


CHANGELOG.md
# CHANGELOG

## [1.0.0] - 2023-09-09

* Initial release


test/aychip.test.js
const { AYChip } = require("..");

test("can synthesise audio in mono", () => {
    const chip = new AYChip({
        frequency: 1773400,
        sampleRate: 11025,
    })
    chip.reset();
    chip.setRegister(0, 8);
    chip.setRegister(1, 0);
    chip.setRegister(2, 8);
    chip.setRegister(3, 0);
    chip.setRegister(4, 8);
    chip.setRegister(5, 0);
    chip.setRegister(6, 8);
    chip.setRegister(7, 0x28);
    chip.setRegister(8, 15);
    chip.setRegister(9, 8);
    chip.setRegister(10, 16);
    chip.setRegister(11, 48);
    chip.setRegister(12, 0);
    chip.setRegister(13, 0x0c);

    buffer = [
        new Float32Array(32),
        new Float32Array(32),
    ];

    chip.generate(buffer, 0, 32);

    resultLeft = Array.from(buffer[0].map(i => Math.floor(i*256)));
    resultRight = Array.from(buffer[0].map(i => Math.floor(i*256)));

    expectedResult = [
        70, 0, 0, 70, 71, 0, 0, 71, 71, 0, 0, 71, 61, 0, 0, 72, 72, 0, 0, 73, 73, 0, 0, 74, 74, 0, 0, 74, 75, 0, 0, 75
    ]

    expect(resultLeft).toStrictEqual(expectedResult);
    expect(resultRight).toStrictEqual(expectedResult);
});

test("can synthesise audio 2", () => {
    const chip = new AYChip({
        frequency: 1773400,
        sampleRate: 11025,
    })
    chip.reset();
    chip.setRegister(0, 8);
    chip.setRegister(1, 0);
    chip.setRegister(2, 8);
    chip.setRegister(3, 0);
    chip.setRegister(4, 8);
    chip.setRegister(5, 0);
    chip.setRegister(6, 8);
    chip.setRegister(7, 0x37);
    chip.setRegister(8, 16);
    chip.setRegister(9, 16);
    chip.setRegister(10, 15);
    chip.setRegister(11, 48);
    chip.setRegister(12, 0);
    chip.setRegister(13, 0x03);

    buffer = [
        new Float32Array(32),
        new Float32Array(32),
    ];

    chip.generate(buffer, 0, 32);

    resultLeft = Array.from(buffer[0].map(i => Math.floor(i*256)));
    resultRight = Array.from(buffer[0].map(i => Math.floor(i*256)));

    expectedResult =     [
        181, 181, 181, 181, 162, 162, 162,
        162, 162, 143, 101, 143, 101, 143,
        129, 129, 129, 129, 129, 114, 114,
         87, 114, 102, 102, 102,  81, 102,
         92,  92,  92,  92
      ]

    expect(resultLeft).toStrictEqual(expectedResult);
    expect(resultRight).toStrictEqual(expectedResult);
});

test("can synthesise audio with stereo mode parameter", () => {
    const chip = new AYChip({
        frequency: 1773400,
        sampleRate: 11025,
        stereoMode: 'acb',
    })
    chip.reset();
    chip.setRegister(0, 8);
    chip.setRegister(1, 0);
    chip.setRegister(7, 0x38);
    chip.setRegister(8, 15);

    buffer = [
        new Float32Array(32),
        new Float32Array(32),
    ];

    chip.generate(buffer, 0, 32);
    expect(buffer[0][0]).toBeCloseTo(0.28866705298423767);
    expect(buffer[0][1]).toBeCloseTo(0);
    expect(buffer[1][0]).toBeCloseTo(0.16666199266910553);
    expect(buffer[1][1]).toBeCloseTo(0);
});

test("can synthesise audio with invalid stereo mode parameter", () => {
    const chip = new AYChip({
        frequency: 1773400,
        sampleRate: 11025,
        stereoMode: 'xyz',
    })
    chip.reset();
    chip.setRegister(0, 8);
    chip.setRegister(1, 0);
    chip.setRegister(7, 0x38);
    chip.setRegister(8, 15);

    buffer = [
        new Float32Array(32),
        new Float32Array(32),
    ];

    chip.generate(buffer, 0, 32);
    expect(buffer[0][0]).toBeCloseTo(0.23569566011428833);
    expect(buffer[0][1]).toBeCloseTo(0);
    expect(buffer[1][0]).toBeCloseTo(0.23569566011428833);
    expect(buffer[1][1]).toBeCloseTo(0);
});

test("can synthesise audio with custom panning parameter and volume", () => {
    const chip = new AYChip({
        frequency: 1773400,
        sampleRate: 11025,
        panning: [0.375, 0.625, 0.5],
        volume: 0.5,
    })
    chip.reset();
    chip.setRegister(0, 8);
    chip.setRegister(1, 0);
    chip.setRegister(7, 0x38);
    chip.setRegister(8, 15);

    buffer = [
        new Float32Array(32),
        new Float32Array(32),
    ];

    chip.generate(buffer, 0, 32);
    expect(buffer[0][0]).toBeCloseTo(0.13175788521766663);
    expect(buffer[0][1]).toBeCloseTo(0);
    expect(buffer[1][0]).toBeCloseTo(0.10205921530723572);
    expect(buffer[1][1]).toBeCloseTo(0);
});


index.js
class AYChip {
    VOLUME_LEVELS = [
        0.000000, 0.004583, 0.006821, 0.009684,
        0.014114, 0.020614, 0.028239, 0.045633,
        0.056376, 0.088220, 0.117568, 0.149977,
        0.190123, 0.229088, 0.282717, 0.333324
    ];

    STEREO_MODES = {
        'ABC': [0.25, 0.5, 0.75],
        'ACB': [0.25, 0.75, 0.5],
        'BAC': [0.5, 0.25, 0.75],
        'BCA': [0.75, 0.25, 0.5],
        'CAB': [0.5, 0.75, 0.25],
        'CBA': [0.75, 0.5, 0.25]
    };

    constructor(opts) {
        this.frequency = opts.frequency;
        this.sampleRate = opts.sampleRate;
        this.cyclesPerSample = this.frequency / this.sampleRate;

        this.volume = ('volume' in opts ? opts.volume : 1);

        if (opts.panning) {
            this.setPanning(opts.panning);
        } else if (opts.stereoMode) {
            this.setPanning(this.STEREO_MODES[opts.stereoMode.toUpperCase()] || [0.5, 0.5, 0.5]);
        } else {
            this.setPanning([0.5, 0.5, 0.5]);
        }

        /* chip state */
        this.toneGeneratorAPhase = 0;
        this.toneGeneratorAPeriod = 8;
        this.toneGeneratorACounter = 0;
    
        this.toneGeneratorBPhase = 0;
        this.toneGeneratorBPeriod = 8;
        this.toneGeneratorBCounter = 0;
    
        this.toneGeneratorCPhase = 0;
        this.toneGeneratorCPeriod = 8;
        this.toneGeneratorCCounter = 0;
    
        this.noiseGeneratorPhase = 0;
        this.noiseGeneratorPeriod = 16;
        this.noiseGeneratorCounter = 0;
        this.noiseGeneratorSeed = 1;
    
        this.toneChanAMask = 0x00;
        this.toneChanBMask = 0x00;
        this.toneChanCMask = 0x00;
        this.noiseChanAMask = 0x00;
        this.noiseChanBMask = 0x00;
        this.noiseChanCMask = 0x00;
    
        this.envelopePeriod = 256;
        this.envelopeCounter = 0;
        this.envelopeRampCounter = 16;
        this.envelopeOnFirstRamp = true;
        this.envelopeAlternateMask = 0x00;
        this.envelopeAlternatePhase = 0x00;
        this.envelopeHoldMask = 0x00;
        this.envelopeAttackMask = 0x00;
        this.envelopeContinueMask = 0x00;
        this.envelopeValue = 0x00;

        this.registers = new Uint8Array(14);
    }

    setPanning(panning) {
        this.panVolumeAdjust = [];
        for (let i = 0; i < 3; i++) {
            /* kebby says we should do this. And you don't argue with kebby.
            http://conspiracy.hu/articles/8/ */
            this.panVolumeAdjust[i] = [
                this.volume * Math.sqrt(1.0-panning[i]), this.volume * Math.sqrt(panning[i])
            ];
        }
    }

    setRegister(reg, val) {
        this.registers[reg] = val;

        switch(reg) {
            case 0:
            case 1:
                this.toneGeneratorAPeriod = (((this.registers[1] & 0x0f) << 8) | this.registers[0]) * 8;
                if (this.toneGeneratorAPeriod === 0) this.toneGeneratorAPeriod = 8;
                break;
            case 2:
            case 3:
                this.toneGeneratorBPeriod = (((this.registers[3] & 0x0f) << 8) | this.registers[2]) * 8;
                if (this.toneGeneratorBPeriod === 0) this.toneGeneratorBPeriod = 8;
                break;
            case 4:
            case 5:
                this.toneGeneratorCPeriod = (((this.registers[5] & 0x0f) << 8) | this.registers[4]) * 8;
                if (this.toneGeneratorCPeriod === 0) this.toneGeneratorCPeriod = 8;
                break;
            case 6:
                this.noiseGeneratorPeriod = (val & 0x1f) * 16;
                if (this.noiseGeneratorPeriod === 0) this.noiseGeneratorPeriod = 16;
                break;
            case 7:
                this.toneChanAMask = (val & 0x01) ? 0xff : 0x00;
                this.toneChanBMask = (val & 0x02) ? 0xff : 0x00;
                this.toneChanCMask = (val & 0x04) ? 0xff : 0x00;
                this.noiseChanAMask = (val & 0x08) ? 0xff : 0x00;
                this.noiseChanBMask = (val & 0x10) ? 0xff : 0x00;
                this.noiseChanCMask = (val & 0x20) ? 0xff : 0x00;
                break;
            case 11:
            case 12:
                this.envelopePeriod = ((this.registers[12] << 8) | this.registers[11]) * 16;
                if (this.envelopePeriod === 0) this.envelopePeriod = 16;
                break;
            case 13:
                this.envelopeCounter = 0;
                this.envelopeRampCounter = 16;
                this.envelopeOnFirstRamp = true;
                this.envelopeAlternatePhase = 0x00;
                this.envelopeHoldMask = (val & 0x01) ? 0x0f : 0x00;
                this.envelopeAlternateMask = (val & 0x02) ? 0x0f : 0x00;
                this.envelopeAttackMask = (val & 0x04) ? 0x0f : 0x00;
                this.envelopeContinueMask = (val & 0x08) ? 0x0f : 0x00;
                break;
        }
    }

    reset() {
        for (let i = 0; i < 14; i++) {
            this.setRegister(i, 0);
        }
    }

    generate(outputBuffer, offset, count) {
        const bufferEnd = offset + count;
        const leftChannelData = outputBuffer[0];
        const rightChannelData = outputBuffer[1];

        for (let bufferPos = offset; bufferPos < bufferEnd; bufferPos++) {

            this.toneGeneratorACounter -= this.cyclesPerSample;
            while (this.toneGeneratorACounter < 0) {
                this.toneGeneratorACounter += this.toneGeneratorAPeriod;
                this.toneGeneratorAPhase ^= 0xff;
            }

            this.toneGeneratorBCounter -= this.cyclesPerSample;
            while (this.toneGeneratorBCounter < 0) {
                this.toneGeneratorBCounter += this.toneGeneratorBPeriod;
                this.toneGeneratorBPhase ^= 0xff;
            }

            this.toneGeneratorCCounter -= this.cyclesPerSample;
            while (this.toneGeneratorCCounter < 0) {
                this.toneGeneratorCCounter += this.toneGeneratorCPeriod;
                this.toneGeneratorCPhase ^= 0xff;
            }

            this.noiseGeneratorCounter -= this.cyclesPerSample;
            while (this.noiseGeneratorCounter < 0) {
                this.noiseGeneratorCounter += this.noiseGeneratorPeriod;

                if ((this.noiseGeneratorSeed + 1) & 2)
                    this.noiseGeneratorPhase ^= 0xff;

                /* rng is 17-bit shift reg, bit 0 is output.
                * input is bit 0 xor bit 3.
                */
                if (this.noiseGeneratorSeed & 1) this.noiseGeneratorSeed ^= 0x24000;
                this.noiseGeneratorSeed >>= 1;
            }

            this.envelopeCounter -= this.cyclesPerSample;
            while (this.envelopeCounter < 0) {
                this.envelopeCounter += this.envelopePeriod;

                this.envelopeRampCounter--;
                if (this.envelopeRampCounter < 0) {
                    this.envelopeRampCounter = 15;
                    this.envelopeOnFirstRamp = false;
                    this.envelopeAlternatePhase ^= 0x0f;
                }

                this.envelopeValue = (
                    /* start with the descending ramp counter */
                    this.envelopeRampCounter
                    /* XOR with the 'alternating' bit if on an even-numbered ramp */
                    ^ (this.envelopeAlternatePhase && this.envelopeAlternateMask)
                );
                /* OR with the 'hold' bit if past the first ramp */
                if (!this.envelopeOnFirstRamp) this.envelopeValue |= this.envelopeHoldMask;
                /* XOR with the 'attack' bit */
                this.envelopeValue ^= this.envelopeAttackMask;
                /* AND with the 'continue' bit if past the first ramp */
                if (!this.envelopeOnFirstRamp) this.envelopeValue &= this.envelopeContinueMask;
            }

            const levelA = this.VOLUME_LEVELS[
                ((this.registers[8] & 0x10) ? this.envelopeValue : (this.registers[8] & 0x0f))
                & (this.toneGeneratorAPhase | this.toneChanAMask)
                & (this.noiseGeneratorPhase | this.noiseChanAMask)
            ];
            const levelB = this.VOLUME_LEVELS[
                ((this.registers[9] & 0x10) ? this.envelopeValue : (this.registers[9] & 0x0f))
                & (this.toneGeneratorBPhase | this.toneChanBMask)
                & (this.noiseGeneratorPhase | this.noiseChanBMask)
            ];
            const levelC = this.VOLUME_LEVELS[
                ((this.registers[10] & 0x10) ? this.envelopeValue : (this.registers[10] & 0x0f))
                & (this.toneGeneratorCPhase | this.toneChanCMask)
                & (this.noiseGeneratorPhase | this.noiseChanCMask)
            ];

            leftChannelData[bufferPos] = (
                this.panVolumeAdjust[0][0] * levelA + this.panVolumeAdjust[1][0] * levelB + this.panVolumeAdjust[2][0] * levelC
            );
            rightChannelData[bufferPos] = (
                this.panVolumeAdjust[0][1] * levelA + this.panVolumeAdjust[1][1] * levelB + this.panVolumeAdjust[2][1] * levelC
            );
        }
    };
}

module.exports = { AYChip };
